iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/static_storage.hpp

```diff
@@ -0,0 +1,108 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_STATIC_STORAGE_HPP
+#define IOX_HOOFS_STATIC_STORAGE_HPP
+
+#include <cstring>
+#include <memory>
+#include <stdint.h>
+
+namespace iox
+{
+namespace cxx
+{
+/// @brief Static storage class to allocate memory for objects of type not yet known.
+///        This storage is not aware of any underlying type.
+///        It can be used where abstract static memory for some object is required.
+///        Currently this memory is allocated on the stack but it could be implemented
+///        to use memory from the static memory segment.
+/// @tparam Capacity number of bytes the static_storage will allocate statically.
+/// @tparam Align alignment of the allocated memory.
+
+/// @note We can define optimized_storage (or dynamic_storage) with a similar interface
+///       but other allocation policies and use them where we need to store objects
+///       with some interchangable storage policy (e.g. in storable_function)
+///       optimized_storage would have a dynamic memory fallback when static memory is
+///       insufficent.
+template <uint64_t Capacity, uint64_t Align = 1>
+class static_storage
+{
+  public:
+    constexpr static_storage() noexcept = default;
+
+    ~static_storage() noexcept;
+
+    /// @note: It is not supposed to be copied or moved for now
+    /// (construct a new one explicitly and populate it instead).
+    /// Move would just be a copy and copy has the problem
+    /// that it would just can perform a memcpy regardless of what is stored
+    /// leading to potentially missed dtor calls
+    static_storage(const static_storage&) = delete;
+    static_storage& operator=(const static_storage&) = delete;
+    static_storage(static_storage&&) = delete;
+    static_storage& operator=(static_storage&&) = delete;
+
+    /// @brief check whether the type T will fit in the buffer 
+    /// @return true if the type fits in the buffer, false otherwise
+    /// @note can be checked at compile time
+    template <typename T>
+    static constexpr bool is_allocatable() noexcept;
+
+    /// @brief provide static memory for an object of type T
+    /// @return pointer to memory where a T can be constructed if memory is available, nullptr otherwise
+    /// @note  compilation fails if storage is insufficient for objects of type T
+    template <typename T>
+    constexpr T* allocate() noexcept;
+
+    /// @brief request aligned memory with a specific size in bytes
+    /// @param align alignment of the memory requested
+    /// @param size number of bytes of the memory requested
+    /// @return pointer to aligned memory of the requested size if available, nullptr otherwise
+    constexpr void* allocate(const uint64_t align, const uint64_t size) noexcept;
```

Comment:
You could think about renaming this method to `aligned_allocate`.

Reply:
No, the parameter reflects this and I do not need the name otherwise. This is not C where we cannot properly overload etc. Function names should be seen combined with the parameters to form something like a "sentence" and redundancy is not required then, if we use proper parameter names (formal and actual) .